txpool/src/containers/dependency.rs

```diff
@@ -0,0 +1,482 @@
+                                    CoinState {
+                                        is_spend_by: Some(tx.id() as TxId),
+                                        depth: 0,
+                                    },
+                                );
+                            } else {
+                                // output out of bound
+                                return Err(Error::NotInsertedOutputNotExisting(*utxo_id).into());
+                            }
+                        } else {
+                            return Err(
+                                Error::NotInsertedInputTxNotExisting(*utxo_id.tx_id()).into()
+                            );
+                        }
+                    }
+
+                    // yey we got our coin
+                }
+                Input::Contract { contract_id, .. } => {
+                    // Does contract exist. We dont need to do any check here other then if contract_id exist or not.
+                    if let Some(state) = self.contracts.get(contract_id) {
+                        // check if contract is created after this transaction.
+                        if state.gas_price > tx.gas_price() {
+                            return Err(Error::NotInsertedContractPricedLower(*contract_id).into());
+                        }
+                        // check depth.
+                        max_depth = core::cmp::max(state.depth, max_depth);
+                        if max_depth > self.max_depth {
+                            return Err(Error::NotInsertedMaxDepth.into());
+                        }
+                    } else {
+                        if !db.contract_exist(*contract_id) {
+                            return Err(
+                                Error::NotInsertedInputContractNotExisting(*contract_id).into()
+                            );
+                        }
+                        // add depth
+                        max_depth = core::cmp::max(1, max_depth);
+                        // insert into contract
+                        db_contracts
+                            .entry(*contract_id)
+                            .or_insert(ContractState {
+                                used_by: HashSet::new(),
+                                depth: 0,
+                                origin: None, //there is no need for owner if contract is in db
+                                gas_price: GasPrice::MAX,
+                            })
+                            .used_by
+                            .insert(tx.id());
+                    }
+
+                    // yey we got our contract
+                }
+            }
+        }
+
+        // now we are sure that transaction can be included. remove all collided transactions
+        let mut removed_tx = Vec::new();
+        for collided in collided.into_iter() {
+            let collided = txs
+                .get(&collided)
+                .expect("Collided should be present in txs");
```

Comment:
Do we really need to panic here?

Reply:
`Collided should be present in txpool` and we already checked that tx is present when comparing `gas_price`.